File: Day01/haskell/src/Main.hs
module Main (main) where

import Data.Char (isSpace)
import Data.List (sort)
import Data.List.Split (splitWhen)

part1 :: [String] -> Int
part1 = head . reverse . sort . map (sum . map read) . splitWhen (all isSpace)

part2 :: [String] -> Int
part2 = sum . take 3 . reverse . sort . map (sum . map read) . splitWhen (all isSpace)

main :: IO ()
main = do
  input <- lines <$> readFile "/home/erik/src/repos/unmango/aoc-2022/Day01/input.txt"
  print (part1 input)
  print (part2 input)

File: Day01/rust/src/main.rs
use std::fs;

fn get_sums(input: &str) -> Vec<u32> {
    return input.split("\n\n")
        .map(|x| x.trim_end()
            .split("\n")
            .map(|y| y.parse::<u32>().expect("To be a u32"))
            .sum::<u32>())
        .collect::<Vec<u32>>();
}

fn part1(input: &str) -> u32 {
    return get_sums(input)
        .into_iter()
        .max()
        .unwrap();
}

fn part2(input: &str) -> u32 {
    let mut sums = get_sums(input);

    sums.sort();

    return sums.into_iter()
        .rev()
        .take(3)
        .sum();
}

fn main() {
    let input: String = fs::read_to_string("../input.txt")
        .expect("Should have read the file");

    let part1_result = part1(&input);
    let part2_result = part2(&input);

    println!("{part1_result:?}");
    println!("{part2_result:?}");
}
